### ISSUE TITLE
[bug-question] Inconsistency between channels and users

### ISSUE THREAD

**Issue by BatuhanK** [centrifugal/centrifugo#359]
**Describe the bug**

Our centrifuge settings allows users to join only their private channel (user#user_id). Also we disabled publish_join and presence settings too.

In theory, when we call HTTP API's with `channels` command we should get online user list.

Here is the problem; when we call this api we get 1000 result (like user#1, user#2, user#3, each unique) but Centrifuge dashboard says `Total clients: 832`.

How this is possible? Which number is correct and shows the current connected users? If total clients (or total users) is correct, is there any way to fetch this users?

![Screen Shot 2020-04-16 at 14 53 27](https://user-images.githubusercontent.com/4688083/79453265-189d5a00-7ff2-11ea-8434-620461bd7256.jpg)

…

**Versions**
> Centrifugo version is 2.4.0
> Client library used is https://github.com/centrifugal/centrifuge-swift 0.0.6
> Operating system is -> using docker image centrifugo/centrifugo:v2.4.0
...

**Comment by FZambia** [comment]
Hi, looking at this picture I can see that some of clients subscribed to several channels. For example on first node you have `199` clients and `246` channels. This means that there are clients with subscriptions to more than 1 channel. Are you sure that you don't have subscriptions to some other channels (I suppose you should see some of them in `channels` API response then)?

Can you reproduce this locally?

**Comment by FZambia** [comment]
Just noticed that `channels` column only updated once a minute - so its value in table does not correspond to `users` and `client` values which get updated once in 3 second at moment. Fixed in https://github.com/centrifugal/centrifuge/commit/23a87fd538e44894f220146ced47a7e946bcf60d

Not sure this will resolve this case though.

**Comment by BatuhanK** [comment]
> Hi, looking at this picture I can see that some of clients subscribed to several channels. For example on first node you have `199` clients and `246` channels. This means that there are clients with subscriptions to more than 1 channel. Are you sure that you don't have subscriptions to some other channels (I suppose you should see some of them in `channels` API response then)?
> 
> Can you reproduce this locally?

I cannot reproduce it. But user's are only connected (and allowed) to their namespaced channel (user#userID).

Is it possible that this api returns channels which are recently abandoned (like user disconnected 2 seconds ago)?

My channels requests and response are below. It returns more than 1200 channnels when it says 1000 clients

```
Request:
{
    "method": "channels",
    "params": {}
}
Response:
{
    "result": {
        "channels": [
            "user#5c52135fb8546700f0eba2a9",
            "user#5e975b9b762920001c8f31fc",
            "user#5c4f77fda72c61003db696a2",
            "user#5e98bc18841494001d9d40b3",
            "user#5cf8c8a2161444002bebcdf1",
            "user#5e979ebd3c5524001bc972c5",
            "user#5e4c05e87172e500231ad4f7",
            "user#5e98a80de49a84001ced4080",
            ...
            "user#59a471e1a5334657d1976801",
            "user#5cdb2c01c5a92c002545e34d"
        ]
    }
}
```

**Comment by BatuhanK** [comment]
I restarted the centrifugo nodes and now it's close to each other (channel ~= client).
But currently (after 15 hours) there is around %15 drift between them... It's frustrating, I don't know why this is happening but I planned to create a global presence based on that :(

**Comment by FZambia** [comment]
`channels` API is not very suitable to rely on it, from [docs](https://centrifugal.github.io/centrifugo/server/http_api/#channels): 

> At moment we mostly suppose that channels command will be used in development process and in not very massive Centrifugo setups (with no more than 10k channels). Also channels command is considered optional in engine implementations.

But anyway it should work correctly.

I think running with fix from my comment above makes some sense - as this can be simply caused by update frequency difference of those counters.

Also you can try to make global channel with presence and compare its output with channels output (by user IDs).

**Comment by BatuhanK** [comment]
> `channels` API is not very suitable to rely on it, from [docs](https://centrifugal.github.io/centrifugo/server/http_api/#channels):
> 
> > At moment we mostly suppose that channels command will be used in development process and in not very massive Centrifugo setups (with no more than 10k channels). Also channels command is considered optional in engine implementations.
> 
> But anyway it should work correctly (with some time approximation though).
> 
> I think running with fix from my comment above makes some sense - as this can be simply caused by update frequency difference of those counters.
> 
> Also you can try to make global channel with presence and compare its output with `channels` output (by user IDs).
> 
> If there is some sort of resource leak then I suppose the difference should grow with time.

Thanks, I guess as you said I will enable presence for single channel and watch changes from it.

**Comment by FZambia** [comment]
@BatuhanK hi, what is the status of this? Have you found sth?

**Comment by BatuhanK** [comment]
> @BatuhanK hi, what is the status of this? Have you found sth?
Currenly no, drift continues dramatically between channels and clients  :/

If you want to examine something on our servers, or is there any dump that I can provide for you I can.

![Screen Shot 2020-04-21 at 11 16 22](https://user-images.githubusercontent.com/4688083/79842779-f88ae380-83c1-11ea-80ef-ba41b65cf8cd.jpg)

**Comment by FZambia** [comment]
Fix from https://github.com/centrifugal/centrifuge/commit/23a87fd538e44894f220146ced47a7e946bcf60d available in v2.5.0
